File: ts/project/.sdk/tm/rust/tests/netsim_test.rs
// Network-behaviour simulation over the offline mock transport (mirrors
// tm/go/test/netsim_test.go). The `test` feature accepts an optional `net`
// config so unit tests can exercise slow, failing and offline conditions
// without a live server. These checks drive the transport through
// direct(), which needs no entity, so they run for every generated SDK
// regardless of its API shape.

use RUSTCRATE::core::helpers::{getp, jo, now_ms};
use RUSTCRATE::{test_sdk, Value};

#[test]
fn netsim_offline_simulation_fails_request() {
    let client = test_sdk(
        jo(vec![("net", jo(vec![("offline", Value::Bool(true))]))]),
        Value::Noval,
    );
    let res = client
        .direct(jo(vec![("path", Value::str("/ping"))]))
        .expect("direct itself should not error");
    assert_eq!(
        getp(&res, "ok"),
        Value::Bool(false),
        "offline network must fail the call"
    );
}

#[test]
fn netsim_failstatus_simulation_surfaces_status() {
    let client = test_sdk(
        jo(vec![(
            "net",
            jo(vec![
                ("failTimes", Value::Num(1.0)),
                ("failStatus", Value::Num(503.0)),
            ]),
        )]),
        Value::Noval,
    );
    let res = client
        .direct(jo(vec![("path", Value::str("/ping"))]))
        .expect("direct itself should not error");
    assert_eq!(getp(&res, "ok"), Value::Bool(false), "expected failed call");
    assert_eq!(
        getp(&res, "status"),
        Value::Num(503.0),
        "expected simulated 503"
    );
}

#[test]
fn netsim_latency_simulation_delays_request() {
    let delay = 60i64;
    let client = test_sdk(
        jo(vec![(
            "net",
            jo(vec![("latency", Value::Num(delay as f64))]),
        )]),
        Value::Noval,
    );
    let start = now_ms();
    client
        .direct(jo(vec![("path", Value::str("/ping"))]))
        .expect("direct itself should not error");
    let elapsed = now_ms() - start;
    // Generous lower bound to stay robust on slow CI.
    assert!(
        elapsed >= delay - 25,
        "expected >= {}ms latency, got {}ms",
        delay - 25,
        elapsed
    );
}

#[test]
fn netsim_plain_test_sdk_works_without_net() {
    let client = test_sdk(Value::Noval, Value::Noval);
    assert_eq!(*client.mode.borrow(), "test");
}

File: ts/project/.sdk/tm/swift/Tests/ProjectNameSDKTests/NetsimTest.swift
// Network-behaviour simulation over the offline mock transport. Drives the
// transport through Direct(), which needs no entity, so these run for every
// generated SDK. Swift twin of the go netsim_test.go.

import XCTest

@testable import ProjectNameSdk

final class NetsimTest: XCTestCase {
  private func directPing(_ net: VMap) -> VMap {
    let testopts = VMap()
    testopts.entries["net"] = .map(net)
    let client = ProjectNameSDK.testSDK(testopts, nil)
    let args = VMap()
    args.entries["path"] = .string("/ping")
    return client.direct(args)
  }

  func testOfflineSimulationFailsRequest() {
    let net = VMap()
    net.entries["offline"] = .bool(true)
    let res = directPing(net)
    XCTAssertEqual(gp(.map(res), "ok"), .bool(false), "offline network must fail the call")
  }

  func testFailstatusSimulationSurfacesStatus() {
    let net = VMap()
    net.entries["failTimes"] = .int(1)
    net.entries["failStatus"] = .int(503)
    let res = directPing(net)
    XCTAssertEqual(gp(.map(res), "ok"), .bool(false), "expected failed call")
    XCTAssertEqual(toInt(gp(.map(res), "status")), 503)
  }

  func testLatencySimulationDelaysRequest() {
    let delay = 60
    let net = VMap()
    net.entries["latency"] = .int(Int64(delay))
    let start = Date()
    _ = directPing(net)
    let elapsedMs = Date().timeIntervalSince(start) * 1000
    XCTAssertGreaterThanOrEqual(elapsedMs, Double(delay - 25),
      "expected >= \(delay - 25)ms latency, got \(elapsedMs)ms")
  }

  func testPlainTestSdkWorksWithoutNet() {
    let client = ProjectNameSDK.testSDK(nil, nil)
    XCTAssertEqual(client.mode, "test")
  }
}
